File: Stack/Leetcode/Medium/MinStack/MinStack.cpp
// https://leetcode.com/problems/min-stack/

#include<bits/stdc++.h>
using namespace std;
class MinStack
{
public:
    stack<pair<int, int>> st;
    MinStack()
    {
    }

    void push(int val)
    {
        int mini = st.empty() || st.top().second > val ? val : st.top().second;
        st.push({val, mini});
    }

    void pop()
    {
        st.pop();
    }

    int top()
    {
        return st.top().first;
    }

    int getMin()
    {
        return st.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Stack/Leetcode/Medium/MinStack/MinStack.java
// https://leetcode.com/problems/min-stack/

import java.util.ArrayDeque;
import java.util.Deque;

public class MinStack {
    class Pair {
        int ele, min;
        Pair(int ele, int min) {
            this.ele = ele;
            this.min = min;
        }
    }
    Deque<Pair> stack;
    public MinStack() {
        stack = new ArrayDeque<>();
    }
    
    public void push(int val) {
        int min = stack.isEmpty() || stack.peek().min > val ? val : stack.peek().min;
        stack.push(new Pair(val, min));
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peek().ele;
    }
    
    public int getMin() {
        return stack.peek().min;
    }
}
